Source/Core/Core/AchievementManager.cpp

```diff
@@ -152,66 +158,218 @@ void AchievementManager::LoadGameByFilenameAsync(const std::string& iso_path,
+    OSD::AddMessage("Achievements are disabled until core is rebooted.", OSD::Duration::VERY_LONG,
+                    OSD::Color::RED);
+    return;
+  }
+  {
+    std::lock_guard lg{m_lock};
+    if (m_loading_volume.get() != nullptr)
     {
-      callback(resolve_hash_response);
-      INFO_LOG_FMT(ACHIEVEMENTS, "No RetroAchievements data found for this game.");
-      OSD::AddMessage("No RetroAchievements data found for this game.", OSD::Duration::VERY_LONG,
-                      OSD::Color::RED);
+      INFO_LOG_FMT(ACHIEVEMENTS, "Disabling Achievement Manager due to hash spam.");
+      DisableRestoreAchievementManager(true);
+      callback(AchievementManager::ResponseType::EXPIRED_CONTEXT);
       return;
     }
-
-    const auto start_session_response = StartRASession();
-    if (start_session_response != ResponseType::SUCCESS)
+    m_loading_volume = DiscIO::CreateVolume(volume->GetBlobReader().CopyReader());
+  }
+  m_system = &Core::System::GetInstance();
+  struct FilereaderState
+  {
+    int64_t position = 0;
+    std::unique_ptr<DiscIO::Volume> volume;
+  };
+  rc_hash_filereader volume_reader{
+      .open = [](const char* path_utf8) -> void* {
+        auto state = std::make_unique<FilereaderState>();
+        {
+          std::lock_guard lg{*AchievementManager::GetInstance()->GetLock()};
+          state->volume = std::move(AchievementManager::GetInstance()->GetLoadingVolume());
+        }
+        if (!state->volume)
+          return nullptr;
+        return state.release();
+      },
+      .seek =
+          [](void* file_handle, int64_t offset, int origin) {
+            switch (origin)
+            {
+            case SEEK_SET:
+              reinterpret_cast<FilereaderState*>(file_handle)->position = offset;
+              break;
+            case SEEK_CUR:
+              reinterpret_cast<FilereaderState*>(file_handle)->position += offset;
+              break;
+            case SEEK_END:
+              // Unused
+              break;
+            }
+          },
+      .tell =
+          [](void* file_handle) {
+            return reinterpret_cast<FilereaderState*>(file_handle)->position;
+          },
+      .read =
+          [](void* file_handle, void* buffer, size_t requested_bytes) {
+            FilereaderState* filereader_state = reinterpret_cast<FilereaderState*>(file_handle);
+            bool success = (filereader_state->volume->Read(
+                filereader_state->position, requested_bytes, reinterpret_cast<u8*>(buffer),
+                DiscIO::PARTITION_NONE));
+            if (success)
+            {
+              filereader_state->position += requested_bytes;
+              return requested_bytes;
+            }
+            else
+            {
+              return static_cast<size_t>(0);
+            }
+          },
+      .close = [](void* file_handle) { delete reinterpret_cast<FilereaderState*>(file_handle); }};
```

Comment:
These are all identical between the two filereaders, please just make them functions instead of lambdas to deduplicate the code here.